src/main/java/com/redhat/devtools/lsp4ij/operations/codelens/LSPCodelensProvider.java

```diff
@@ -21,91 +21,177 @@
-        CodeLensParams param = new CodeLensParams(new TextDocumentIdentifier(fileUri.toASCIIString()));
-        BlockingDeque<Pair<CodeLens, LanguageServer>> pairs = new LinkedBlockingDeque<>();
-
-        CompletableFuture<Void> future = collect(file, project, param, pairs, cancellationSupport);
-        List<Pair<Integer, Pair<CodeLens, LanguageServer>>> codeLenses = createCodeLenses(document, pairs, future, cancellationSupport);
-        codeLenses.stream()
-                .collect(Collectors.groupingBy(p -> p.first))
-                .forEach((offset, list) ->
-                        inlayHintsSink.addBlockElement(
-                                offset,
-                                true,
-                                true,
-                                0,
-                                toPresentation(editor, offset, list, factory, cancellationSupport))
-                );
+    protected void doCollect(@NotNull VirtualFile file,
+                             @NotNull PsiFile psiFile,
+                             @NotNull Project project,
+                             @NotNull Editor editor,
+                             @NotNull PresentationFactory factory,
+                             @NotNull InlayHintsSink inlayHintsSink,
+                             @NotNull CancellationSupport cancellationSupport) throws InterruptedException {
+        // Get from cache code lenses or create them
+        List<CompletableFuture> resolvedCodeLensFutures = new ArrayList<>();
+        List<Pair<Integer, CodeLensData>> codeLenses = Collections.emptyList();
+        try {
+            codeLenses = getOrCreateCodeLenses(file, psiFile, project, editor, cancellationSupport);
+
+
+            // Render code lenses and collect all codeLens/resolve futures which are not finished
+            codeLenses.stream()
+                    .collect(Collectors.groupingBy(p -> p.first))
+                    .forEach((offset, list) ->
+                            inlayHintsSink.addBlockElement(
+                                    offset,
+                                    true,
+                                    true,
+                                    0,
+                                    toPresentation(editor, offset, list, factory, resolvedCodeLensFutures, cancellationSupport))
+                    );
+
+        }
+        finally {
+            CodeLensCache cache = editor.getUserData(CACHE);
+            CompletableFuture<List<CodeLensData>> future = cache.getCodeLenses(psiFile.getModificationStamp());
+            if (!future.isDone()) {
+                resolvedCodeLensFutures.add(future);
+            }
+            if (!resolvedCodeLensFutures.isEmpty() && !cancellationSupport.isCanceled()) {
+                final List<Pair<Integer, CodeLensData>> lenses = codeLenses;
+                // There are some codeLens/resolve which are not finished, wait for all finish of those futures and refresh the inlay hints
```

Comment:
Some codeLens/resolve requests are not finished, wait for all of those futures to complete and refresh the inlay hints to render them